### PR TITLE
Propagate feature discovery errors (#543)

### LINKED ISSUE #543: Unit Architecture: propagate errors from find_feature_files instead of swallowing them
## Summary

`find_feature_files` is a public query function returning `Vec<PathBuf>`, but `std::fs::read_dir` and directory-iteration errors are silently swallowed via `let Ok(...) else { return; }` and `filter_map(Result::ok)`. This means a filesystem error (e.g., permission denied, broken symlink) causes the function to return an empty `Vec` rather than signalling failure, which can produce silent false-negatives in test discovery.

## Resolution

- Change `find_feature_files` (and any helper functions) to return `Result<Vec<PathBuf>, ServerError>` (or an equivalent error type).
- Propagate `read_dir` failures and iteration errors instead of discarding them.
- Update all call sites to handle the error at the appropriate boundary.

## References

- Raised during review of PR #519 (https://github.com/leynos/rstest-bdd/pull/519)
- Requested by @leynos

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch makes `find_feature_files` and its directory traversal helpers return `Result` so workspace, recursive directory and directory-entry failures reach callers as `ServerError::Io` instead of producing silent false-negatives. It updates the public API example and adapts all in-repository callers to handle the fallible result.

Closes #543.

## Review walkthrough

- Review [`crates/rstest-bdd-server/src/discovery/workspace.rs`](https://github.com/leynos/rstest-bdd/blob/e49487a8bd114d46814d2cb98068532e2f029935/crates/rstest-bdd-server/src/discovery/workspace.rs) for the public return-type change, error propagation through both traversal helpers, caller updates and portable regression tests for workspace-root and recursive-directory read failures.

## Validation

- `make check-fmt`: passed.
- `make test`: passed; 1,658 Rust tests passed, 7 were skipped, and 69 Python tests passed.
- `make typecheck`: passed.
- `make lint`: passed, including Clippy, Whitaker, Ruff, Pylint and repository contract checks.
- `git diff --check`: passed.
- `coderabbit review --agent`: passed with zero findings after the implementation, final portability adjustment and committed milestone.

## References

- Issue: [Unit Architecture: propagate errors from find_feature_files instead of swallowing them #543](https://github.com/leynos/rstest-bdd/issues/543)
- Review context: [Fold v0.6.0-beta2 GPUI adopter feedback into design, roadmap, and ADRs #519](https://github.com/leynos/rstest-bdd/pull/519)
- [Lody session](https://lody.ai/leynos/sessions/88046ccf-b6f6-46de-bb6a-8e939f1d4996)

## Summary by Sourcery

Propagate filesystem errors from workspace feature-file discovery to callers.

New Features:
- Make workspace feature-file discovery return a fallible result that reports filesystem traversal failures.

Bug Fixes:
- Prevent workspace, recursive-directory, metadata, and directory-entry errors from being silently treated as missing feature files.

Enhancements:
- Update callers and public examples to propagate discovery errors instead of producing incomplete or empty results.

Documentation:
- Document the fallible discovery API, error-handling guidance, and migration requirements.

Tests:
- Add unit, property-based, behavioral, and compile-time coverage for successful discovery and propagated I/O failures.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[crates/rstest-bdd-server/src/discovery/workspace.rs](https://github.com/leynos/rstest-bdd/pull/627/files/e49487a8bd114d46814d2cb98068532e2f029935#diff-cfeb70588e03f16c24b2f9fe75081bce1fc27a5c311aa34e11caabdddf9d59b1)

Comment on lines +334 to +348

~~~rust
    fn reports_recursive_directory_read_failure() {
        let missing_directory = TempDir::new().expect("failed to create temp dir");
        let missing_path = missing_directory.path().to_path_buf();
        missing_directory
            .close()
            .expect("failed to remove temp dir");
        let mut features = Vec::new();

        let error = collect_feature_files_recursive(&missing_path, &mut features)
            .expect_err("missing feature directory should return an I/O error");

        assert!(
            matches!(error, ServerError::Io(ref source) if source.kind() == io::ErrorKind::NotFound)
        );
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=627&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-server%2Fsrc%2Fdiscovery%2Fworkspace.rs&method=)
The module contains 2 functions with similar structure: tests.reports_recursive_directory_read_failure,tests.reports_workspace_read_failure

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
crates/rstest-bdd-server/src/discovery/workspace.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: tests.reports_recursive_directory_read_failure,tests.reports_workspace_read_failure

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: crates/rstest-bdd-server/src/discovery/workspace.rs [hunk 1]
```diff
@@ -105,19 +105,23 @@ fn find_manifest_path(path: &Path) -> Result<PathBuf, ServerError> {
 ///
 /// * `workspace_root` - The root directory of the workspace
 ///
+/// # Errors
+///
+/// Returns `ServerError::Io` when the workspace or a feature directory cannot
+/// be read, including failures encountered while iterating directory entries.
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/src/discovery/workspace.rs:114]: @coderabbitai Fixed in 5a4a1b0. Feature-directory and Cargo.toml probes now use fallible metadata checks. Optional locations ignore only NotFound; other metadata failures propagate as ServerError::Io. The private reader seam supplies deterministic metadata and directory-entry failure regression tests.

**Comment by leynos** [comment]
@coderabbitai Verified the previous pre-merge findings against 5a4a1b0:

- Fixed the metadata-probe error swallowing and added deterministic metadata and nested directory-entry error tests.
- The existing parameterized public-path test already covers successful nested recursive discovery; the new reader seam covers the otherwise non-deterministic iterator-failure boundary.
- Added migration and developer guidance. The users guide is not applicable because this is a library-facing LSP helper, not user-facing CLI behaviour.
- Did not add property/proof or trybuild coverage: OS filesystem errors are runtime effects best covered by deterministic reader tests, and trybuild targets compiler diagnostics rather than this runtime API contract.

Validated with make check-fmt, make test, make typecheck, make lint, and make markdownlint.

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[crates/rstest-bdd-server/src/discovery/workspace_tests.rs](https://github.com/leynos/rstest-bdd/pull/627/files/5a4a1b00bf30372dcd0bddafb770cb4830596773#diff-bf736bdb9f7984f116261df717a6b1673ab24bb6531cd89197d6c722b5cfa815)

Comment on lines +220 to +231

~~~rust
fn reports_recursive_directory_entry_failure() {
    let reader = FailingDirectoryReader {
        failure: ReaderFailure::NestedDirectoryEntry,
    };
    let mut features = Vec::new();

    let error =
        collect_feature_files_recursive(&reader, Path::new("feature-directory"), &mut features)
            .expect_err("unreadable nested feature directory should return an I/O error");

    assert_io_error_kind(error, io::ErrorKind::PermissionDenied);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=627&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-server%2Fsrc%2Fdiscovery%2Fworkspace_tests.rs&method=)
The module contains 2 functions with similar structure: reports_nested_feature_directory_entry_failure,reports_recursive_directory_entry_failure

File: crates/rstest-bdd-server/src/discovery/workspace.rs [hunk 2]
```diff
@@ -128,26 +132,28 @@ pub fn find_feature_files(workspace_root: &Path) -> Vec<PathBuf> {
 
     for dir in &search_dirs {
         if dir.is_dir() {
-            collect_feature_files_recursive(dir, &mut features);
+            collect_feature_files_recursive(dir, &mut features)?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/src/discovery/workspace.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd-server/src/discovery/workspace_tests.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    assert_eq!(info.root, workspace.path());
+}
+
+#[rstest]
+fn fails_when_no_manifest_found() {
+    let dir = TempDir::new().expect("failed to create temp dir");
+    let result = discover_workspace(dir.path());
+    assert!(result.is_err());
+    let err = result.unwrap_err();
+    assert!(err.to_string().contains("no Cargo.toml found"));
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/src/discovery/workspace_tests.rs:65]: @coderabbitai Fixed in 90b19292. The test now uses `expect_err` with the explicit message that workspace discovery should fail when `Cargo.toml` is absent.

File: crates/rstest-bdd-server/src/discovery/workspace.rs [hunk 4]
```diff
@@ -127,174 +176,85 @@ pub fn find_feature_files(workspace_root: &Path) -> Vec<PathBuf> {
...
-    fn returns_empty_when_no_feature_files(create_test_workspace: io::Result<TempDir>) {
-        let features = find_feature_files(
-            create_test_workspace
-                .expect("test setup should succeed")
-                .path(),
-        );
-        assert!(features.is_empty());
-    }
-}
+mod tests;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/src/discovery/workspace.rs:272]: @coderabbitai Fixed in 90b19292. The broad module-level `unwrap_used` expectation was removed. The updated tests use `expect_err`, and no unwrap-like calls remain that require a narrow expectation.

**Comment by leynos** [comment]
@coderabbitai Verified on 90b19292:

- Fixed the remaining deterministic reader-seam coverage: manifest metadata `PermissionDenied` and workspace-root directory-entry iteration errors now return `ServerError::Io`.
- Added the user-facing workspace discovery API contract to the users guide; the migration and developer guidance are also present.
- Skipped a separate integration test: the existing discovery test module already calls the exported `find_feature_files` API and covers successful nested discovery plus I/O failure at the public boundary, so a second crate-level wrapper would add no distinct behaviour.
- Skipped trybuild: this is a runtime API result-handling contract, not a compiler-diagnostic interface; the normal workspace compilation and test gates already compile its callers.

Validated after the final rebase with `make check-fmt`, `make test`, `make typecheck`, `make lint`, and `make markdownlint`.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (1 error, 4 warnings)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Testing (Overall) | ❌ Error | FAIL — The tests cover public result handling, missing workspace paths, recursive read failures, and one metadata and entry failure path. They do not guard all changed error paths. `is_optional_file` … | Add deterministic `DirectoryReader` tests that make `is_file` return `PermissionDenied` for a crate manifest and assert `ServerError::Io`. Add a test that makes the workspace-root entry iterator return an error and assert that `search_crate… |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 53.57% which is insufficient. The required threshold is 80.00%. Docstring coverage is scoped to functions touched by this diff. Analyzed 28 functions across 2 files. (2 skipped: … | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| User-Facing Documentation | ⚠️ Warning | The pull request changes the public `rstest_bdd_server::discovery::find_feature_files` API from `Vec<PathBuf>` to `Result<Vec<PathBuf>, ServerError>`, and the module re-exports it for API consumers. T… | Add a user-facing section to `docs/users-guide.md`, preferably under `Language server`, for API consumers of `rstest-bdd-server`. Document the new `Result<Vec<PathBuf>, ServerError>` signature, the propagation of workspace and directory tra… |
| Testing (Unit And Behavioural) | ⚠️ Warning | Add behavioural coverage at the public API boundary. The pull request changes the exported `discovery::find_feature_files` contract from `Vec<PathBuf>` to `Result<Vec<PathBuf>, ServerError>`, but it a… | Add an integration test under `crates/rstest-bdd-server/tests` that imports `rstest_bdd_server::discovery::find_feature_files` and verifies successful discovery and `ServerError::Io` at the public boundary. Keep the private seam tests for d… |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Fail the compile-time testing check. The PR changes the public Rust API from `pub fn find_feature_files(&Path) -> Vec<PathBuf>` to `pub fn find_feature_files(&Path) -> Result<Vec, ServerError… | Add a Rust trybuild test for `rstest-bdd-server`. Include a compile-pass fixture that handles `Result<Vec<PathBuf>, ServerError>` and a compile-fail fixture that uses the old `Vec<PathBuf>` contract, with a checked-in `.stderr` expectation.… |

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (4 warnings)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Testing (Unit And Behavioural) | ⚠️ Warning | Add a behavioural test at the public `find_feature_files` boundary. The PR changes an exported API and documents its `ServerError::Io` contract, but the recursive read and directory-entry failure test… | Add an integration or other public-boundary test that imports `rstest_bdd_server::discovery::find_feature_files` and verifies the fallible discovery contract for recursive traversal failures, not only the private helper. Use a portable dete… |
| Testing (Property / Proof) | ⚠️ Warning | The change introduces a universal traversal invariant over a range of filesystem layouts and operation outcomes: every non-optional I/O failure from `read_dir`, directory-entry iteration, or metadata … | Add substantive `proptest` coverage for the discovery contract. Generate bounded directory trees and bounded sequences of injected failures across workspace reads, recursive reads, directory-entry iteration, `is_directory`, and `is_file`. A… |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Fail the check because the PR changes the exported Rust API from `Vec<PathBuf>` to `Result<Vec<PathBuf>, ServerError>`, which introduces compile-time caller behaviour. The PR adds only runtime unit te… | Add a `trybuild` or equivalent Rust compile-time test for the public `find_feature_files` contract. Pin a valid caller that handles `Result<Vec<PathBuf>, ServerError>` and, where the breaking API is part of the contract, pin the rejected di… |
| Observability | ⚠️ Warning | The pull request introduces a new externally visible filesystem-failure outcome in the public `find_feature_files` API. The changed traversal now returns `ServerError::Io` from `read_dir`, directory-e… | Add observability at the public feature-discovery boundary. Instrument `find_feature_files` with a span named for the operation and record elapsed time. Emit a warning or error when the operation returns a non-optional I/O failure, with sta… |

**Comment by leynos** [comment]
@coderabbitai Please investigate the cause of the following issue using codegraph exploration and research, identify a fix and provide an AI coding agent prompt for the fix:

~~~
         Compiling vte v0.14.1
  
      (test timed out)
  
    Cancelling due to test failure: 1 test still running
          PASS [ 102.659s] (1713/1717) rstest-bdd::trybuild_macros step_macros_compile
  ────────────
       Summary [ 112.343s] 1713/1717 tests run: 1712 passed, 1 timed out, 7 skipped
       TIMEOUT [  60.006s] (1712/1717) rstest-bdd-server::workspace_discovery_compile find_feature_files_requires_result_handling
  warning: 4/1717 tests were not run due to test failure (run with --no-fail-fast to run all tests, or run with --max-fail)
  error: test run failed
  error: process didn't exit successfully: `/home/runner/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/cargo nextest run --manifest-path Cargo.toml --target-dir /home/runner/work/rstest-bdd/rstest-bdd/target/llvm-cov-target --workspace --no-default-features --features 'rstest-bdd/diagnostics rstest-bdd-server/test-support strict-compile-time-validation'` (exit status: 100)
  cargo llvm-cov nextest --manifest-path Cargo.toml --workspace --no-default-features --features 'rstest-bdd/diagnostics rstest-bdd-server/test-support strict-compile-time-validation' --cobertura --output-path coverage.rust.xml failed with code 1
~~~

**Comment by wafflecat-df12** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (3 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Testing (Overall) | ❌ Error | Strengthen the observability tests before merge. The filesystem tests are substantive: they assert successful and empty discovery, public `NotFound` handling, metadata failures, workspace and recursiv… | Add a tracing subscriber test layer that records new spans and span-field updates. Assert that successful and failed calls create the `feature_file_discovery` span with the expected `workspace_root` and a finite, non-negative `elapsed_secon… |
| Unit Architecture | ❌ Error | The fallible filesystem query is explicit and propagates `ServerError::Io`, but the changed public query now performs observability side-effects. `find_feature_files` directly calls `describe_counter!… | Move metrics, tracing, warning emission, and elapsed-time measurement out of the public `find_feature_files` query. Keep the query focused on directory reads and its explicit `Result<Vec<PathBuf>, ServerError>` contract. Place observability… |
| Rust Compiler Lint Integrity | ❌ Error | Remove the artificial function-item references in `find_feature_files_result.rs:18` and `find_feature_files_vec.rs:14`. Each `let _ = discover_features;` only marks an otherwise unused helper as used.… | Replace each function-item reference with a real invocation from `main`, such as `let _ = discover_features(Path::new("."));`, or remove the references and update the compile-fail expectation to retain the compiler's unused-function diagnos… |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 72.73% which is insufficient. The required threshold is 80.00%. Docstring coverage is scoped to functions touched by this diff. Analyzed 55 functions across 7 files. (3 skipped: … | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Testing (Property / Proof) | ⚠️ Warning | The change introduces a quantified traversal invariant: for arbitrary workspace trees, recursion depth, entry order, and failure location, non-optional directory, metadata, and iterator errors must re… | Add a substantive `proptest` suite for discovery using a bounded in-memory `DirectoryReader` model. Generate directory trees, nested depths, entry orderings, feature-file leaves, and injected `read_dir`, iterator, and metadata failures. Ass… |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[crates/rstest-bdd-server/src/discovery/workspace_failure_tests.rs](https://github.com/leynos/rstest-bdd/pull/627/files/a875c4c154e6280f79aa3349c849e505543582fa#diff-2bdc45117f3e7d446bbf74a44b1dda120cadb7759d77b86639bd216ceef6b2da)

Comment on lines +121 to +132

~~~rust
fn reports_recursive_directory_entry_failure() {
    let reader = FailingDirectoryReader {
        failure: ReaderFailure::NestedDirectoryEntry,
    };
    let mut features = Vec::new();

    let error =
        collect_feature_files_recursive(&reader, Path::new("feature-directory"), &mut features)
            .expect_err("unreadable nested feature directory should return an I/O error");

    assert_io_error_kind(error, io::ErrorKind::PermissionDenied);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=627&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-server%2Fsrc%2Fdiscovery%2Fworkspace_failure_tests.rs&method=)
The module contains 2 functions with similar structure: reports_nested_feature_directory_entry_failure,reports_recursive_directory_entry_failure

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### docs/developers-guide.md (1)

`200-211`: _📐 Maintainability & Code Quality_ | _🟡 Minor_ | _⚡ Quick win_
**Update the trybuild timeout documentation.**
Add `rstest-bdd-server::workspace_discovery_compile` to this override list and

change the binary count. The PR adds this compile-test binary to the

10-minute override, but this text still documents only three binaries.
Triage: `[type:docstyle]`
As per coding guidelines, “New or changed ... tooling or build requirements

must be clearly documented in the developer's guide” and the guide must stay

synchronised with the repository.

##### 🤖 Detailed instructions

Use a wyvern agent team to verify each finding against the current code and only fix it if needed. Fix any documentation issues using a scribe agent team. Fix only still-valid issues, skip the rest with a brief reason, keep changes minimal, and validate. Use a scrutineer agent for execution of tests and summarizing the results.

In `@docs/developers-guide.md` around lines 200 - 211, Update the trybuild-based
compile-test timeout documentation to include
rstest-bdd-server::workspace_discovery_compile and revise the wording from three
binaries to four, keeping the existing 10-minute timeout and cargo-spawning
serialization details accurate.

_Sources: Coding guidelines, Path instructions_

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[crates/rstest-bdd-server/src/discovery/workspace_tests.rs](https://github.com/leynos/rstest-bdd/pull/627/files/85f7a8e446ab15788363dba650d6689cfbf6494e#diff-bf736bdb9f7984f116261df717a6b1673ab24bb6531cd89197d6c722b5cfa815)

Comment on lines +187 to +228

~~~rust
    fn read_dir(&self, path: &Path) -> io::Result<Self::Entries> {
        match self.failure {
            Some((FailureSite::WorkspaceReadDirectory, kind)) if path == Path::new("workspace") => {
                return Err(io::Error::from(kind));
            }
            Some((FailureSite::RecursiveDirectoryRead, kind))
                if path == Path::new(RECURSIVE_FAILURE_DIRECTORY) =>
            {
                return Err(io::Error::from(kind));
            }
            _ => {}
        }
        let mut entries = self.entries.get(path).cloned().unwrap_or_default();
        let entry_failure_kind = match self.failure {
            Some((FailureSite::WorkspaceDirectoryEntry, kind))
                if path == Path::new("workspace") =>
            {
                Some(kind)
            }
            Some((FailureSite::RecursiveDirectoryEntry, kind))
                if path == Path::new(RECURSIVE_FAILURE_DIRECTORY) =>
            {
                Some(kind)
            }
            _ => None,
        };
        if entry_failure_kind.is_some() {
            entries.insert(0, PathBuf::new());
        }
        Ok(entries
            .into_iter()
            .map(move |entry| {
                if entry.as_os_str().is_empty()
                    && let Some(kind) = entry_failure_kind
                {
                    return Err(io::Error::from(kind));
                }
                Ok(entry)
            })
            .collect::<Vec<_>>()
            .into_iter())
    }
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=627&biomarker=Bumpy+Road+Ahead&filename=crates%2Frstest-bdd-server%2Fsrc%2Fdiscovery%2Fworkspace_tests.rs&method=InMemoryDirectoryReader.read_dir)
InMemoryDirectoryReader.read_dir has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[crates/rstest-bdd-server/src/discovery/workspace_tests.rs](https://github.com/leynos/rstest-bdd/pull/627/files/85f7a8e446ab15788363dba650d6689cfbf6494e#diff-bf736bdb9f7984f116261df717a6b1673ab24bb6531cd89197d6c722b5cfa815)

Comment on lines +187 to +228

~~~rust
    fn read_dir(&self, path: &Path) -> io::Result<Self::Entries> {
        match self.failure {
            Some((FailureSite::WorkspaceReadDirectory, kind)) if path == Path::new("workspace") => {
                return Err(io::Error::from(kind));
            }
            Some((FailureSite::RecursiveDirectoryRead, kind))
                if path == Path::new(RECURSIVE_FAILURE_DIRECTORY) =>
            {
                return Err(io::Error::from(kind));
            }
            _ => {}
        }
        let mut entries = self.entries.get(path).cloned().unwrap_or_default();
        let entry_failure_kind = match self.failure {
            Some((FailureSite::WorkspaceDirectoryEntry, kind))
                if path == Path::new("workspace") =>
            {
                Some(kind)
            }
            Some((FailureSite::RecursiveDirectoryEntry, kind))
                if path == Path::new(RECURSIVE_FAILURE_DIRECTORY) =>
            {
                Some(kind)
            }
            _ => None,
        };
        if entry_failure_kind.is_some() {
            entries.insert(0, PathBuf::new());
        }
        Ok(entries
            .into_iter()
            .map(move |entry| {
                if entry.as_os_str().is_empty()
                    && let Some(kind) = entry_failure_kind
                {
                    return Err(io::Error::from(kind));
                }
                Ok(entry)
            })
            .collect::<Vec<_>>()
            .into_iter())
    }
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=627&biomarker=Complex+Method&filename=crates%2Frstest-bdd-server%2Fsrc%2Fdiscovery%2Fworkspace_tests.rs&method=InMemoryDirectoryReader.read_dir)
InMemoryDirectoryReader.read_dir has a cyclomatic complexity of 10, threshold = 9
